File: src/charting/renderer/XAxisRendererHorizontalBarChart.java
package com.github.mikephil.charting.renderer;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Paint.Align;
import android.graphics.Path;
import android.graphics.RectF;

import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.components.LimitLine;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.XAxis.XAxisPosition;
import com.github.mikephil.charting.utils.FSize;
import com.github.mikephil.charting.utils.MPPointF;
import com.github.mikephil.charting.utils.MPPointD;
import com.github.mikephil.charting.utils.Transformer;
import com.github.mikephil.charting.utils.Utils;
import com.github.mikephil.charting.utils.ViewPortHandler;

import java.util.List;

public class XAxisRendererHorizontalBarChart extends XAxisRenderer {

    protected BarChart this.mChart;

    public XAxisRendererHorizontalBarChart(viewPortHandler: ViewPortHandler, XAxis xAxis,
            Transformer trans, BarChart chart) {
        super(viewPortHandler, xAxis, trans);

        this.mChart = chart;
    }

    
    public computeAxis(let min, let max, boolean inverted) {

        // calculate the starting and entry polet of the y-labels (depending on
        // zoom / contentrect bounds)
        if (mViewPortHandler.contentWidth() > 10 && !mViewPortHandler.isFullyZoomedOutY()) {

            MPPointD p1 = this.mTrans.getValuesByTouchPoint(mViewPortHandler.contentLeft(), this.mViewPortHandler.contentBottom());
            MPPointD p2 = this.mTrans.getValuesByTouchPoint(mViewPortHandler.contentLeft(), this.mViewPortHandler.contentTop());

            if (inverted) {

                min =  p2.y;
                max =  p1.y;
            } else {

                min =  p1.y;
                max =  p2.y;
            }

            MPPointD.recycleInstance(p1);
            MPPointD.recycleInstance(p2);
        }

        computeAxisValues(min, max);
    }

    
    protected computeSize() {

        this.mAxisLabelPaint.setTypeface(mXAxis.getTypeface());
        this.mAxisLabelPaint.setTextSize(mXAxis.getTextSize());

        let longest = this.mXAxis.getLongestLabel();

        final FSize labelSize = Utils.calcTextSize(mAxisLabelPaint, longest);

        const labelWidth = (labelSize.width + this.mXAxis.getXOffset() * 3.5f);
        const labelHeight = labelSize.height;

        final FSize labelRotatedSize = Utils.getSizeOfRotatedRectangleByDegrees(
                labelSize.width,
                labelHeight,
                this.mXAxis.getLabelRotationAngle());

        this.mXAxis.mLabelWidth = Math.round(labelWidth);
        this.mXAxis.mLabelHeight = Math.round(labelHeight);
        this.mXAxis.mLabelRotatedWidth = (labelRotatedSize.width + this.mXAxis.getXOffset() * 3.5f);
        this.mXAxis.mLabelRotatedHeight = Math.round(labelRotatedSize.height);

        FSize.recycleInstance(labelRotatedSize);
    }

    
    public renderAxisLabels(c: Canvas) {

        if (!mXAxis.isEnabled() || !mXAxis.isDrawLabelsEnabled())
            return;

        let xoffset = this.mXAxis.getXOffset();

        this.mAxisLabelPaint.setTypeface(mXAxis.getTypeface());
        this.mAxisLabelPaint.setTextSize(mXAxis.getTextSize());
        this.mAxisLabelPaint.setColor(mXAxis.getTextColor());

        MPPointF pointF = MPPointF.getInstance(0,0);

        if (mXAxis.getPosition() == XAxisPosition.TOP) {
            pointF.x = 0.0;
            pointF.y = 0.5f;
            drawLabels(c, this.mViewPortHandler.contentRight() + xoffset, pointF);

        } else if (mXAxis.getPosition() == XAxisPosition.TOP_INSIDE) {
            pointF.x = 1.0;
            pointF.y = 0.5f;
            drawLabels(c, this.mViewPortHandler.contentRight() - xoffset, pointF);

        } else if (mXAxis.getPosition() == XAxisPosition.BOTTOM) {
            pointF.x = 1.0;
            pointF.y = 0.5f;
            drawLabels(c, this.mViewPortHandler.contentLeft() - xoffset, pointF);

        } else if (mXAxis.getPosition() == XAxisPosition.BOTTOM_INSIDE) {
            pointF.x = 1.0;
            pointF.y = 0.5f;
            drawLabels(c, this.mViewPortHandler.contentLeft() + xoffset, pointF);

        } else { // BOTH SIDED
            pointF.x = 0.0;
            pointF.y = 0.5f;
            drawLabels(c, this.mViewPortHandler.contentRight() + xoffset, pointF);
            pointF.x = 1.0;
            pointF.y = 0.5f;
            drawLabels(c, this.mViewPortHandler.contentLeft() - xoffset, pointF);
        }

        MPPointF.recycleInstance(pointF);
    }

    
    protected drawLabels(c: Canvas, let pos, MPPointF anchor) {

        const labelRotationAngleDegrees = this.mXAxis.getLabelRotationAngle();
        boolean centeringEnabled = this.mXAxis.isCenterAxisLabelsEnabled();

        float[] positions = new float[mXAxis.mEntryCount * 2];

        for (let i = 0; i < positions.length; i += 2) {

            // only fill x values
            if (centeringEnabled) {
                positions[i + 1] = this.mXAxis.mCenteredEntries[i / 2];
            } else {
                positions[i + 1] = this.mXAxis.mEntries[i / 2];
            }
        }

        this.mTrans.pointValuesToPixel(positions);

        for (let i = 0; i < positions.length; i += 2) {

            let y = positions[i + 1];

            if (mViewPortHandler.isInBoundsY(y)) {

                let label = this.mXAxis.getValueFormatter().getAxisLabel(mXAxis.mEntries[i / 2], this.mXAxis);
                drawLabel(c, label, pos, y, anchor, labelRotationAngleDegrees);
            }
        }
    }

    
    public RectF getGridClippingRect() {
        this.mGridClippingRect.set(mViewPortHandler.getContentRect());
        this.mGridClippingRect.inset(0, -mAxis.getGridLineWidth());
        return this.mGridClippingRect;
    }

    
    protected drawGridLine(c: Canvas, let x, let y, Path gridLinePath) {

        gridLinePath.moveTo(mViewPortHandler.contentRight(), y);
        gridLinePath.lineTo(mViewPortHandler.contentLeft(), y);

        // draw a path because lines don't support dashing on lower android versions
        c.drawPath(gridLinePath, this.mGridPaint);

        gridLinePath.reset();
    }

    
    public renderAxisLine(c: Canvas) {

        if (!mXAxis.isDrawAxisLineEnabled() || !mXAxis.isEnabled())
            return;

        this.mAxisLinePaint.setColor(mXAxis.getAxisLineColor());
        this.mAxisLinePaint.setStrokeWidth(mXAxis.getAxisLineWidth());

        if (mXAxis.getPosition() == XAxisPosition.TOP
                || this.mXAxis.getPosition() == XAxisPosition.TOP_INSIDE
                || this.mXAxis.getPosition() == XAxisPosition.BOTH_SIDED) {
            c.drawLine(mViewPortHandler.contentRight(),
                    this.mViewPortHandler.contentTop(), this.mViewPortHandler.contentRight(),
                    this.mViewPortHandler.contentBottom(), this.mAxisLinePaint);
        }

        if (mXAxis.getPosition() == XAxisPosition.BOTTOM
                || this.mXAxis.getPosition() == XAxisPosition.BOTTOM_INSIDE
                || this.mXAxis.getPosition() == XAxisPosition.BOTH_SIDED) {
            c.drawLine(mViewPortHandler.contentLeft(),
                    this.mViewPortHandler.contentTop(), this.mViewPortHandler.contentLeft(),
                    this.mViewPortHandler.contentBottom(), this.mAxisLinePaint);
        }
    }

    protected Path this.mRenderLimitLinesPathBuffer = new Path();
    /**
	 * Draws the LimitLines associated with this axis to the screen.
	 * This is the standard YAxis renderer using the XAxis limit lines.
	 *
	 * @param c
	 */
	
	public renderLimitLines(c: Canvas) {

		List<LimitLine> limitLines = this.mXAxis.getLimitLines();

		if (limitLines == null || limitLines.length <= 0)
			return;

		float[] pts = this.mRenderLimitLinesBuffer;
        pts[0] = 0;
        pts[1] = 0;

		Path limitLinePath = this.mRenderLimitLinesPathBuffer;
        limitLinePath.reset();

		for (let i = 0; i < limitLines.length; i++) {

			LimitLine l = limitLines.get(i);

            if(!l.isEnabled())
                continue;

            let clipRestoreCount = c.save();
            this.mLimitLineClippingRect.set(mViewPortHandler.getContentRect());
            this.mLimitLineClippingRect.inset(0, -l.getLineWidth());
            c.clipRect(mLimitLineClippingRect);

			this.mLimitLinePaint.setStyle(Paint.Style.STROKE);
			this.mLimitLinePaint.setColor(l.getLineColor());
			this.mLimitLinePaint.setStrokeWidth(l.getLineWidth());
			this.mLimitLinePaint.setPathEffect(l.getDashPathEffect());

			pts[1] = l.getLimit();

			this.mTrans.pointValuesToPixel(pts);

			limitLinePath.moveTo(mViewPortHandler.contentLeft(), pts[1]);
			limitLinePath.lineTo(mViewPortHandler.contentRight(), pts[1]);

			c.drawPath(limitLinePath, this.mLimitLinePaint);
			limitLinePath.reset();
			// c.drawLines(pts, this.mLimitLinePaint);

			let label = l.getLabel();

			// if drawing the limit-value label is enabled
			if (label != null && !label.equals("")) {

				this.mLimitLinePaint.setStyle(l.getTextStyle());
				this.mLimitLinePaint.setPathEffect(null);
				this.mLimitLinePaint.setColor(l.getTextColor());
				this.mLimitLinePaint.setStrokeWidth(0.5f);
				this.mLimitLinePaint.setTextSize(l.getTextSize());

                const labelLineHeight = Utils.calcTextHeight(mLimitLinePaint, label);
                let xOffset = Utils.convertDpToPixel(4f) + l.getXOffset();
                let yOffset = l.getLineWidth() + labelLineHeight + l.getYOffset();

                final LimitLine.LimitLabelPosition position = l.getLabelPosition();

				if (position == LimitLine.LimitLabelPosition.RIGHT_TOP) {

					this.mLimitLinePaint.setTextAlign(Align.RIGHT);
					c.drawText(label,
                            this.mViewPortHandler.contentRight() - xOffset,
							pts[1] - yOffset + labelLineHeight, this.mLimitLinePaint);

				} else if (position == LimitLine.LimitLabelPosition.RIGHT_BOTTOM) {

                    this.mLimitLinePaint.setTextAlign(Align.RIGHT);
                    c.drawText(label,
                            this.mViewPortHandler.contentRight() - xOffset,
                            pts[1] + yOffset, this.mLimitLinePaint);

                } else if (position == LimitLine.LimitLabelPosition.LEFT_TOP) {

                    this.mLimitLinePaint.setTextAlign(Align.LEFT);
                    c.drawText(label,
                            this.mViewPortHandler.contentLeft() + xOffset,
                            pts[1] - yOffset + labelLineHeight, this.mLimitLinePaint);

                } else {

					this.mLimitLinePaint.setTextAlign(Align.LEFT);
					c.drawText(label,
                            this.mViewPortHandler.offsetLeft() + xOffset,
							pts[1] + yOffset, this.mLimitLinePaint);
				}
			}

            c.restoreToCount(clipRestoreCount);
		}
	}
}

File: src/charting/renderer/XAxisRendererHorizontalBarChart.ts
import { Align, Canvas, Paint, Path, RectF } from '@nativescript-community/ui-canvas';
import { profile } from '@nativescript/core';
import { BarChart } from '../charts/BarChart';
import { LimitLabelPosition } from '../components/LimitLine';
import { XAxis, XAxisPosition } from '../components/XAxis';
import { MPPointF } from '../utils/MPPointF';
import { Transformer } from '../utils/Transformer';
import { Utils } from '../utils/Utils';
import { ViewPortHandler } from '../utils/ViewPortHandler';
import { CustomRendererGridLineFunction } from './AxisRenderer';
import { XAxisRenderer } from './XAxisRenderer';

export class XAxisRendererHorizontalBarChart extends XAxisRenderer {
    protected mChart: BarChart;
    mForceLongestLabelComputation = true;

    constructor(viewPortHandler: ViewPortHandler, xAxis: XAxis, trans: Transformer, chart: BarChart) {
        super(viewPortHandler, xAxis, trans);
        this.mChart = chart;
    }

    public computeAxis(min, max, inverted) {
        // calculate the starting and entry polet of the y-labels (depending on
        // zoom / contentrect bounds)
        if (this.mViewPortHandler.getContentRect().width() > 10 && !this.mViewPortHandler.isFullyZoomedOutY()) {
            const rect = this.mAxis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();
            const p1 = this.mTrans.getValuesByTouchPoint(rect.left, rect.bottom);
            const p2 = this.mTrans.getValuesByTouchPoint(rect.left, rect.top);

            if (inverted) {
                min = p2.y;
                max = p1.y;
            } else {
                min = p1.y;
                max = p2.y;
            }
        }

        this.computeAxisValues(min, max);
    }

    protected computeSize() {
        const axis = this.mXAxis;
        const longest = axis.getLongestLabel();
        const paint = this.axisLabelsPaint;
        paint.setFont(axis.getFont());

        const labelSize = Utils.calcTextSize(paint, longest);

        const labelWidth = labelSize.width + axis.getXOffset() * 3.5;
        const labelHeight = labelSize.height;

        const labelRotatedSize = Utils.getSizeOfRotatedRectangleByDegrees(labelSize.width, labelHeight, axis.getLabelRotationAngle());

        axis.mLabelWidth = Math.round(labelWidth);
        axis.mLabelHeight = Math.round(labelHeight);
        axis.mLabelRotatedWidth = labelRotatedSize.width + axis.getXOffset() * 3.5;
        axis.mLabelRotatedHeight = Math.round(labelRotatedSize.height);
    }

    @profile
    public renderAxisLabels(c: Canvas) {
        const axis = this.mXAxis;
        if (!axis.isEnabled() || !axis.isDrawLabelsEnabled()) {
            return;
        }

        const xOffset = axis.getXOffset();

        const paint = this.axisLabelsPaint;
        paint.setFont(axis.getFont());
        paint.setTextAlign(axis.getLabelTextAlign());
        paint.setColor(axis.getTextColor());

        const pointF = { x: 0, y: 0 };

        const rect = this.mAxis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();
        if (axis.getPosition() === XAxisPosition.TOP) {
            pointF.x = 0.0;
            pointF.y = 0.5;
            this.drawLabels(c, rect.right + xOffset, pointF);
        } else if (axis.getPosition() === XAxisPosition.TOP_INSIDE) {
            pointF.x = 1.0;
            pointF.y = 0.5;
            this.drawLabels(c, rect.right - xOffset, pointF);
        } else if (axis.getPosition() === XAxisPosition.BOTTOM) {
            pointF.x = 1.0;
            pointF.y = 0.5;
            this.drawLabels(c, rect.left - xOffset, pointF);
        } else if (axis.getPosition() === XAxisPosition.BOTTOM_INSIDE) {
            pointF.x = 1.0;
            pointF.y = 0.5;
            this.drawLabels(c, rect.left + xOffset, pointF);
        } else {
            // BOTH SIDED
            pointF.x = 0.0;
            pointF.y = 0.5;
            this.drawLabels(c, rect.right + xOffset, pointF);
            pointF.x = 1.0;
            pointF.y = 0.5;
            this.drawLabels(c, rect.left - xOffset, pointF);
        }
    }

    /**
     * Draws the x-labels on the specified y-position.
     *
     * @param pos
     */
    @profile
    protected drawLabels(c: Canvas, pos, anchor: MPPointF) {
        const axis = this.mXAxis;
        const labelRotationAngleDegrees = axis.getLabelRotationAngle();
        const centeringEnabled = axis.isCenterAxisLabelsEnabled();
        const entryCount = axis.mEntryCount;
        if (entryCount === 0) {
            return;
        }
        if (!this.mLabelsPositionsBuffer || this.mLabelsPositionsBuffer.length !== length) {
            this.mLabelsPositionsBuffer = Utils.createArrayBuffer(length);
        }
        const positionsBuffer = this.mLabelsPositionsBuffer;
        for (let i = 0; i < length; i += 2) {
            // only fill x values
            if (centeringEnabled) {
                positionsBuffer[i] = axis.mCenteredEntries[i / 2];
            } else {
                positionsBuffer[i] = axis.mEntries[i / 2];
            }
            if (i + 1 < length) {
                positionsBuffer[i + 1] = 0;
            }
        }
        const positions = Utils.pointsFromBuffer(positionsBuffer);

        this.mTrans.pointValuesToPixel(positions);
        const labels = axis.mLabels;
        const paint = this.axisLabelsPaint;
        for (let i = 0; i < positions.length; i += 2) {
            const y = positions[i + 1];
            const label = labels[i / 2];
            if (!label) {
                continue;
            }
            if (this.mViewPortHandler.isInBoundsY(y)) {
                this.drawLabel(c, label, pos, y, anchor, labelRotationAngleDegrees, paint);
            }
        }
    }

    protected drawGridLine(c: Canvas, rect: RectF, x, y, axisValue, paint: Paint, customRendererFunc: CustomRendererGridLineFunction) {
        if (customRendererFunc) {
            customRendererFunc(c, this, rect, x, y, axisValue, paint);
        } else {
            c.drawLine(rect.right, y, rect.left, y, paint);
        }
    }

    public renderAxisLine(c: Canvas) {
        const axis = this.mXAxis;
        if (!axis.isDrawAxisLineEnabled() || !axis.isEnabled()) {
            return;
        }
        const paint = this.axisLinePaint;

        paint.setColor(axis.getAxisLineColor());
        paint.setStrokeWidth(axis.getAxisLineWidth());
        const rect = this.mAxis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();

        if (axis.getPosition() === XAxisPosition.TOP || axis.getPosition() === XAxisPosition.TOP_INSIDE || axis.getPosition() === XAxisPosition.BOTH_SIDED) {
            c.drawLine(rect.right, rect.top, rect.right, rect.bottom, paint);
        }

        if (axis.getPosition() === XAxisPosition.BOTTOM || axis.getPosition() === XAxisPosition.BOTTOM_INSIDE || axis.getPosition() === XAxisPosition.BOTH_SIDED) {
            c.drawLine(rect.left, rect.top, rect.left, rect.bottom, paint);
        }
    }

    /**
     * Draws the LimitLines associated with this axis to the screen.
     * This is the standard YAxis renderer using the XAxis limit lines.
     *
     * @param c
     */

    public renderLimitLines(c: Canvas) {
        const axis = this.mXAxis;
        const limitLines = axis.getLimitLines();

        if (limitLines == null || limitLines.length <= 0) return;

        const pts = Utils.getTempArray(2);
        pts[0] = 0;
        pts[1] = 0;

        const limitLinePath = Utils.getTempPath();
        limitLinePath.reset();
        let offsetLeft = 0;
        let rect: RectF;
        if (axis.isIgnoringOffsets()) {
            rect = this.mViewPortHandler.getChartRect();
        } else {
            rect = this.mViewPortHandler.getContentRect();
            offsetLeft = this.mViewPortHandler.offsetLeft();
        }
        const customRender = axis.getCustomRenderer();
        const customRendererFunc = customRender && customRender.drawLimitLine;
        const clipToContent = axis.clipLimitLinesToContent;
        for (let i = 0; i < limitLines.length; i++) {
            const l = limitLines[i];

            if (!l.isEnabled()) {
                continue;
            }
            const lineWidth = l.getLineWidth();
            if (clipToContent) {
                const clipRect = Utils.getTempRectF();
                clipRect.set(rect);
                clipRect.inset(0, -lineWidth);
                c.clipRect(clipRect);
            }

            const paint = this.limitLinePaint;
            paint.setColor(l.getLineColor());
            paint.setStrokeWidth(lineWidth);
            paint.setPathEffect(l.getDashPathEffect());

            pts[1] = l.getLimit();

            this.mTrans.pointValuesToPixel(pts);

            if (lineWidth > 0) {
                if (customRendererFunc) {
                    customRendererFunc(c, this, l, rect, pts[1], paint);
                } else {
                    c.drawLine(rect.left, pts[1], rect.right, pts[1], paint);
                }
            }

            limitLinePath.lineTo(rect.right, pts[1]);

            c.drawPath(limitLinePath, paint);
            limitLinePath.reset();

            const label = l.getLabel();

            // if drawing the limit-value label is enabled
            if (label != null && label !== '') {
                paint.setStyle(l.getTextStyle());
                paint.setPathEffect(null);
                paint.setColor(l.getTextColor());
                paint.setStrokeWidth(0.5);
                paint.setFont(l.getFont());

                const labelLineHeight = Utils.calcTextHeight(paint, label);
                const xOffset = 4 + l.getXOffset();
                const yOffset = l.getLineWidth() + labelLineHeight + l.getYOffset();

                const position = l.getLabelPosition();

                switch (position) {
                    case LimitLabelPosition.RIGHT_TOP: {
                        paint.setTextAlign(Align.RIGHT);
                        c.drawText(label, rect.right - xOffset, pts[1] - yOffset + labelLineHeight, paint);
                        break;
                    }
                    case LimitLabelPosition.RIGHT_BOTTOM: {
                        paint.setTextAlign(Align.RIGHT);
                        c.drawText(label, rect.right - xOffset, pts[1] + yOffset, paint);
                        break;
                    }
                    case LimitLabelPosition.CENTER_TOP: {
                        paint.setTextAlign(Align.CENTER);
                        c.drawText(label, rect.right, pts[1] - yOffset + labelLineHeight, paint);
                        break;
                    }
                    case LimitLabelPosition.CENTER_BOTTOM: {
                        paint.setTextAlign(Align.CENTER);
                        c.drawText(label, rect.right, pts[1] + yOffset, paint);
                        break;
                    }
                    case LimitLabelPosition.LEFT_TOP: {
                        paint.setTextAlign(Align.LEFT);
                        c.drawText(label, rect.left + xOffset, pts[1] - yOffset + labelLineHeight, paint);
                        break;
                    }
                    case LimitLabelPosition.LEFT_BOTTOM: {
                        paint.setTextAlign(Align.LEFT);
                        c.drawText(label, offsetLeft + xOffset, pts[1] + yOffset, paint);
                        break;
                    }
                }
            }

            if (clipToContent) {
                c.restore();
            }
        }
    }
}
